sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -15,6 +15,7 @@ namespace Azure.Core
     public sealed class HttpMessage : IDisposable
     {
         private Dictionary<string, object>? _properties;
+        private Dictionary<Type, object>? _typeProperties;
```

Comment:
Wouldn't this be a breaking change?

Reply:
Well, I meant we would reimplement the existing public method that write and read the string-object dictionary to use the new store, i.e. 
```csharp
SetProperty(string key, object value){
    Dictionary<string, object> properties;
    if (!_typeProperties.TryGet(typeof(PrivateType), out properties){
        properties = new Dictionary<string, object>();
        properties.Add(key, value);
        _typeProperties.Add(typeof(PrivateType), properties);
    }
    properties[key] = value;
}
```

Reply:
Thinking more about this, it's probably worth while to do something like this to avoid the additional field, which grows the allocation size of every `HttpMessage` by 8 bytes.